### PR TITLE
ILU device support without UVM

### PR DESCRIPTION
**Pull request by victorapm**
Expanded from #792. This PR has three main goals:

- Enable ILU to work without UVM on device builds.
- Add the runtime switch option to ILU. Thus, in device builds, one can choose now whether to execute ILU's setup and solve on the device or host.
- Improve error handling. e.g, when trying to execute on the device ILU types that are not supported yet.

Also, ILU0's performance on the device has been improved, e.g., for a 27-pt 3D Poisson test case, ILU setup times on a V100 card are give below
| Size        | master 53dfbe360f582b7d23f56a56c36e6b120f9f165f | Ilu-uvm 16c186798fbfc045ef5c299affac904a80b25a0c|
| ----------- | ------ | ------- |
| 64x64x64    | 0.394  | 0.296   |
| 92x92x92    | 1.194  | 0.951   |
| 128x128x128 | 3.408  | 2.773   |
| 160x160x160 | 6.956  | 5.632   |
| 192x192x192 | 12.551 | 10.15   |

Note: this PR depends on #815.

### CODE REVIEW

**Review by rfalgout — APPROVED** [review]
Looks good to me.  Quite a bit of work, too!  I noticed a few things that are being added through separate pull requests, but I assume your plan is to synchronize appropriately when the time comes.  Thanks!

**Comment by victorapm** [comment]
Thanks, Rob! Yes, this PR depends on #815. So I need to merge 815 into master first.

**Comment by victorapm** [comment]
@liruipeng @oseikuffuor1 Some additional info: regression tests with this branch are passing as of [4fed272](https://github.com/hypre-space/hypre/pull/816/commits/4fed2728774b3b38e2d1dab6560077afd7b49b4f)

**Review by ulrikeyang — APPROVED** [review]
This looks good, however I did not look very carefully over some of the large files, since I am not that familiar with that code than some of you are, and I don't want to hold this up.

File: src/parcsr_ls/par_ilu_setup.c [hunk 1]
```diff
@@ -407,311 +398,494 @@ hypre_ILUSetup( void               *ilu_vdata,
...
+         if (exec == HYPRE_EXEC_DEVICE)
          {
-            /* Only support ILU0 */
-            hypre_ILUSetupILU0Device(matA, perm, perm, n, nLU, matL_des, matU_des, ilu_solve_policy,
-                                     &ilu_solve_buffer,
-                                     &matBL_info, &matBU_info, &matSL_info, &matSU_info, &matBLU_d, &matS,
-                                     &matE_d, &matF_d, &A_diag_fake, 1);//BJ + cusparse_ilu0()
+            if (fill_level == 0)
+            {
+               /* BJ + cusparse_ilu0() - Only support ILU0 */
```
Comment (oseikuffuor1) [inline src/parcsr_ls/par_ilu_setup.c]: Can you change the comment here to read GMRES + ... (instead of BJ)

File: src/parcsr_ls/par_ilu_setup.c [hunk 2]
```diff
@@ -407,311 +398,494 @@ hypre_ILUSetup( void               *ilu_vdata,
...
+            }
+            else
+            {
+#if !defined(HYPRE_USING_UNIFIED_MEMORY)
+               hypre_error_w_msg(HYPRE_ERROR_GENERIC,
+                                 "GMRES+ILUK setup on device runs requires unified memory!");
+               return hypre_error_flag;
+#endif
+
+               /* BJ + cusparse_ilu0() */
```
Comment (oseikuffuor1) [inline src/parcsr_ls/par_ilu_setup.c]: Same comment as above. GMRES + ILUK. A few more below for the ILUT variants as well.

File: src/parcsr_ls/par_ilu_setup.c (hunk 1 above)
Comment (victorapm) [reply src/parcsr_ls/par_ilu_setup.c]: Thanks for catching! Done in [7c78be2](https://github.com/hypre-space/hypre/pull/816/commits/7c78be2eb456fc781ea7af9e9e9e8a07f5681484)

File: src/parcsr_ls/par_ilu_setup.c (hunk 2 above)
Comment (victorapm) [reply src/parcsr_ls/par_ilu_setup.c]: Done in [7c78be2](https://github.com/hypre-space/hypre/pull/816/commits/7c78be2eb456fc781ea7af9e9e9e8a07f5681484)

**Review by oseikuffuor1 — APPROVED** [review]
Thanks @victorapm !! This is a lot of good work!

File: src/parcsr_ls/par_ilu_setup.c [hunk 3]
```diff
@@ -1232,58 +1432,10 @@ hypre_ILUSetup( void               *ilu_vdata,
...
-/* Extract submatrix from diagonal part of A into a new CSRMatrix without sort rows
- * WARNING: We don't put diagonal to the first entry of each row since this function is now for cuSparse only
- * A = input matrix
- * perm = permutation array indicating ordering of rows. Perm could come from a
- *    CF_marker array or a reordering routine.
- * rqperm = reverse permutation array indicating ordering of columns
- * A_diagp = pointer to the output diagonal matrix.
- */
-HYPRE_Int
-hypre_ParILUCusparseExtractDiagonalCSR( hypre_ParCSRMatrix *A,
```
Comment (liruipeng) [inline src/parcsr_ls/par_ilu_setup.c:1244]: So, was this function removed?

Comment (victorapm) [reply src/parcsr_ls/par_ilu_setup.c:1244]: Yes, `hypre_CSRMatrixPermute` is doing its job

File: src/parcsr_ls/par_ilu.c [hunk 4]
```diff
@@ -2296,173 +2423,187 @@ hypre_ILULocalRCM( hypre_CSRMatrix *A, HYPRE_Int start, HYPRE_Int end,
...
-               }
+               HYPRE_Int tmp = G_capacity;
+               G_capacity = (HYPRE_Int) (G_capacity * EXPAND_FACT + 1);
+               G_j = hypre_TReAlloc_v2(G_j, HYPRE_Int, tmp, HYPRE_Int,
+                                       G_capacity, HYPRE_MEMORY_HOST);
             }
          }
       }
-      G_i[num_nodes] = G_nnz;
-      if (G_nnz == 0)
```
Comment (liruipeng) [inline src/parcsr_ls/par_ilu.c:2360]: I am not sure about this part of the code, which seems to be gone. Was it unnecessary?

**Review by liruipeng — APPROVED** [review]

File: src/parcsr_ls/par_ilu.c (hunk 4 above)
Comment (victorapm) [reply src/parcsr_ls/par_ilu.c:2360]: In the new code, when `G_nnz == 0`, I am still freeing the variables `G_j`, `rqperm` and `G`. But that happens later on. I reorganized the logic to have less lines of code

Comment (liruipeng) [reply src/parcsr_ls/par_ilu.c:2360]: OK. How about the else branch, like `A + A'` ? Is it gone as well?

Comment (victorapm) [reply src/parcsr_ls/par_ilu.c:2360]: Yes, I reorganized the code so that the `else` block from `if (sym)` is taken into account as well

Comment (liruipeng) [reply src/parcsr_ls/par_ilu.c:2360]: Ok . I guess it’s fine as long as the regression is good. I couldn’t follow all the details.

Comment (liruipeng) [reply src/parcsr_ls/par_ilu.c:2360]: Ok . I guess it’s fine as long as the regression is good. I couldn’t follow all the details.

**Pull request merged**